tools/mo/openvino/tools/mo/front/kaldi/add_reshape_around_convolution.py

```diff
@@ -1,21 +1,109 @@
+        if in_node.time_dim == -1:
+            return None
+        if in_node.time_dim > in_frame_time_max:
+            in_frame_time_max = in_node.time_dim
+
+    return in_frame_time_max
+
+
+def propagate_time_dim_through_branch(node: Node, dest_port: Port):
+    """
+    Propagate time_dim for one branch starting from given node:dest_port.
+    MemoryOffset/Splice nodes increase time_dim
+    Convolution/Pooling decrease time_dim to 0 value because it works through the whole context
+    other nodes don't change time_dim
+    If we find out node with several inputs, one of which have undefined time_dim, process stops.
+    """
+    child = dest_port.node
+
+    if child.time_dim >= 0:
+        return
+
+    if child.op == "MemoryOffset":
+        child.time_dim = node.time_dim + abs(child.t)
+    elif child.op == "Splice":
+        child.time_dim = node.time_dim + len(child.context) - 1
+    elif child.op in ['Convolution', 'Pooling']:
+        child.time_dim = 0
+    elif len(child.in_edges()) > 1:
+        tmp_time_dim = find_max_frame_time(child)
+        if tmp_time_dim is not None:
+            child.time_dim = tmp_time_dim
+        else:
+            return
+    else:  # len(child.in_edges()) == 1
+        child.time_dim = node.time_dim
+
+    for p in child.out_ports():
+        if child.out_port(p).disconnected():
+            continue
+        for dest in child.out_port(p).get_destinations():
+            propagate_time_dim_through_branch(child, dest)
+
+
+def set_time_dim(graph):
+    """
+    Set value of dimension where we gather frames with different time labels.
+    If in some node we don't use any context, then time_dim = 0
+    """
+    nx.set_node_attributes(G=graph, name='time_dim', values=-1)
+    start_nodes = graph.get_op_nodes(op="Const")
+    start_nodes.extend(graph.get_op_nodes(op='Parameter'))
+    for n in start_nodes:
+        n.time_dim = 0
+
+    for node in start_nodes:
+        for p in node.out_ports():
+            if node.out_port(p).disconnected():
+                continue
+            for dest in node.out_port(p).get_destinations():
+                propagate_time_dim_through_branch(node, dest)
+
+
+def update_time_dim_for_start_convolution(graph):
+    """
+    If we have pattern like Parameter->Convolution->... then input already spliced outside network. So from set_time_dim
+    time_dim will be set as 1 and it will be wrong. For such pattern time_dim should be set as kernel[1]
+    (convolution run through the whole splice)
+    """
+    params = graph.get_op_nodes(op="Parameter")
+    for param_node in params:
+        if not param_node.out_port(0).disconnected() and \
+                param_node.out_port(0).get_destination().node.op == 'Convolution':
+            conv_node = param_node.out_port(0).get_destination().node
+            param_node.time_dim = conv_node.soft_get('kernel')[1] - 1
+
+
 class ReplaceConvolutionReshape(FrontReplacementPattern):
```

Comment:
I think class and file names should be updated. Now only convolution was mentioned in the name while indeed it also add Reshapes and Transposes for Pools as well.

Also Transposes are not mentioned in the names

Reply:
done